executor/aggregate.go

```diff
@@ -45,39 +96,482 @@ type HashAggExec struct {
+	// Init partial workers.
+	for i := 0; i < partialConcurrency; i++ {
+		baseAggWorker := baseHashAggWorker{
+			finishCh:     e.finishCh,
+			aggFuncs:     e.AggFuncs,
+			maxChunkSize: e.maxChunkSize,
+			groupVals:    make([][]byte, 0, 8),
+		}
+
+		w := HashAggPartialWorker{
+			baseHashAggWorker: baseAggWorker,
+			inputCh:           make(chan *chunk.Chunk, 1),
+			outputChs:         e.partialOutputChs,
+			giveBackCh:        e.inputCh,
+			globalOutputCh:    e.finalOutputCh,
+			aggCtxsMap:        make(aggCtxsMapper, 0),
+			groupByItems:      e.GroupByItems,
+			chk:               e.children[0].newChunk(),
+		}
+
+		e.partialWorkers[i] = w
+		e.inputCh <- &HashAggInput{
+			chk:        e.children[0].newChunk(),
+			giveBackCh: w.inputCh,
+		}
+	}
+
+	// Init final workers.
+	finalAggFuncs := e.newFinalAggFuncs()
+	for i := 0; i < finalConcurrency; i++ {
+		baseAggWorker := baseHashAggWorker{
+			finishCh:     e.finishCh,
+			aggFuncs:     finalAggFuncs,
+			maxChunkSize: e.maxChunkSize,
+			groupVals:    make([][]byte, 0, 8),
+		}
+		e.finalWorkers[i] = HashAggFinalWorker{
+			baseHashAggWorker:   baseAggWorker,
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupSet:            mvmap.NewMVMap(),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup, finalConcurrency int) {
+	needShuffle, sc := false, ctx.GetSessionVars().StmtCtx
+	defer func() {
+		if needShuffle {
+			w.shuffleIntermData(sc, finalConcurrency)
+		}
+		waitGroup.Done()
+	}()
+	for {
+		select {
+		case <-w.finishCh:
+			return
+		case chk, ok := <-w.inputCh:
+			if !ok {
+				return
+			}
+			w.chk.SwapColumns(chk)
+			w.giveBackCh <- &HashAggInput{
+				chk:        chk,
+				giveBackCh: w.inputCh,
+			}
+		}
+
+		if err := w.updateIntermData(ctx, sc, w.chk, len(w.aggCtxsMap)); err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return
+		}
+		needShuffle = true
+	}
+}
+
+func (w *HashAggPartialWorker) updateIntermData(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (err error) {
```

Comment:
`updatePartialResult` is a better function name?

Reply:
IntermData seems to be more accurate?
When we support `distinct`, `PartialResult` will be somewhat ambiguous.

Reply:
once we supported the `distinct`, and I think maybe it can takes us a long time, we can change it back.